Implementation of Datahose loop

symphony-soufiane opened the pull request:
### Description
Implemented Datahose loop. It is based on an ack_id like Datafeed V2, that is why a refactor is done in order to get the common logic in a new abstract class abstract_ackId_event_loop, that in turn extends abstract_datafeed_loop and defines common interface.
Datahose loop is not supposed to be used to react to activities but it is meant to subscribe to different real time events for monitoring and scanning purposes, that is why the interface does not allow to register activities.

The generated api/code has been updated to support last changes.

Closes #280

### Checklist
- [x] Referenced an issue in the PR title or description
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [x] Updated the documentation in [docsrc folder](../tree/main/docsrc)

symphony-soufiane committed:
Implementation of Datahose loop.

Implemented Datahose loop. It is based on an ack_id like Datafeed V2, that is why a refactor is done in order to get the common logic in a new abstract class abstract_ackId_event_loop, that in turn extends abstract_datafeed_loop and defines common intereface.
Datahose loop is not supposed to be used to react to activities but it is meant to subscribe to different real time events for monitoring and scanning purposes, that is why the interface does not allow to register activities.

The generated api/code has been updated to support last changes.

symphony-soufiane committed:
fix broken unit tests and clean code

symphony-soufiane committed:
Introduced abstract_datahose_loop interface to distinguish datahose from datafeed implementation from a developer point of view, given both are instances of abtract_datafeed_loop

symphony-youri commented on api_client_generation/generate.sh:
```diff
@@ -4,7 +4,7 @@ code_gen_dir=`pwd`
 echo $code_gen_dir
 
-commit_hash=2f23eb7bff43da0743f372e15cc754190e7a9a63
+commit_hash=3171d0d55e1c1e4d1dd4931a91bf78c61d29f4d8
```
Please make sure it is 76f5817b321c055048e3b4caeedbc0c3d3f86890 now (with the PR I just merged)

symphony-elias commented on symphony/bdk/core/config/model/bdk_datahose_config.py:
```diff
@@ -0,0 +1,24 @@
+        self.filters = None
+        self.retry = BdkRetryConfig(dict(maxAttempts=BdkRetryConfig.INFINITE_MAX_ATTEMPTS))
+        if config is not None:
+            self.tag = config.get(TAG)
```
We should set a default value for tag and filters if config is None (maybe None value).

symphony-soufiane replied:
Isn't it the case here? we first initialise tag and filters with None and "retry" with default one, and then we override them if config is not None.
The default tag (datahose-BOT_NAME) or (datahose) is set in DatahoseLoop constuctor 

https://github.com/finos/symphony-bdk-python/pull/281/files/33baa811ecb26d83f22300c833b94e43e46ae4a3#diff-c036be8e109b086eabf605d69cf2d381e5375847f1d086bb4942784f529fe247R36

symphony-elias commented on symphony/bdk/core/service/datafeed/abstract_datafeed_loop.py:
```diff
@@ -86,26 +87,28 @@ def __init__(self, datafeed_api: DatafeedApi, session_service: SessionService, a
 
         :param hard_kill: if set to True, tasks running listener methods will be cancelled immediately. Otherwise, tasks
-          will be awaited until completion.
+          will be waited until completion.
```
`tasks will be waited for` or `tasks will be awaited`

symphony-elias commented on symphony/bdk/core/service/datafeed/datafeed_loop_v2.py:
```diff
@@ -1,35 +1,33 @@
 
 # Based on the DFv2 default visibility timeout, after which an event is re-queued
-EVENT_PROCESSING_MAX_DURATION_SECONDS = 30
+# EVENT_PROCESSING_MAX_DURATION_SECONDS = 30
```
Please remove this line

symphony-elias commented on symphony/bdk/core/service/datafeed/abstract_ackId_event_loop.py:
```diff
@@ -0,0 +1,75 @@
+        super().__init__(datafeed_api, session_service, auth_session, config)
+        self._ack_id = ""
+
+    async def recreate_datafeed(self):
```
This method is specific to DatafeedLoopV2, it should be removed

symphony-soufiane replied:
**_recreate_datafeed()_** and **_prepare_datafeed()_** are both specific to **_DatafeedLoopV2_** (and V1 too*) 
We have them here because **_abstract_ackId_event_loop_** extends **_abstract_datafeed_loop_** where both methods are declared. 
The interest of extending **_abstract_datafeed_loop_** is to reuse the same code as most of it is used for **DF1, DF2 and Datahose loops.** 

Options I can see:
Keep both methods declared and not implemented in DatahoseLoop.
Remove this inheritance and duplicate the code.
Introduce one more interface (abstract_recreative_datafeed_loop) and we move both methods there. (this should not break backward compatibility IMO since the methods are aimed to only be used internally)

symphony-elias replied:
@symphony-soufiane I would prefer to have these method duplicated in DatafeedLoopV1 and DatafeedLoopV2 as it is quite specific to each version. And it is weird to have an empty "prepare_datafeed" method in the DatahoseLoop class

symphony-soufiane replied:
Done in https://github.com/finos/symphony-bdk-python/pull/281/commits/f17597d4fee7a3c5c1f22afe7af066bb13073f49

symphony-soufiane committed:
added unit tests

symphony-soufiane committed:
Merge branch 'main' into support-datahose

symphony-soufiane committed:
Excluded datahose feeds or invalid feed ids when reusing datafeeds in DFv2

symphony-soufiane committed:
Merge branch 'support-datahose' of https://github.com/symphony-soufiane/symphony-api-client-python into support-datahose

symphony-soufiane committed:
Update unit test having datahose loop with invalid config session.
Datahose constructor sets the default tag as a concatenation of Datahose keyword and the bot name. A bot should be configured to use Datahose. An exception is thrown thanks to bot_service annotation in symphony_bdk, but we still need to check if the bot name is not None in Datahose loop as it is executed earlier, and it ends with a different  error when no bot is configured.

symphony-soufiane committed:
moved common fixtures in a dedicated file

symphony-soufiane committed:
added docs

symphony-soufiane committed:
Corrected link to activity-api

symphony-elias commented on docsrc/markdown/datafeed.md:
```diff
@@ -155,6 +155,109 @@ datafeed:
+difference that *all* events of the pod are received. The datahose loop is a core service built on top of the events API
+and provide a dedicated contract to bot developers to work with datahose. This is compatible with agent version 20.15 onwards.
+
+The [Activity API](../activity-api.md) is not meant to be used with datahose.
```
Have you tried if the link works by generating the documentation manually (`make html`) ? I guess there should be only one dot instead of two.

symphony-elias commented on docsrc/markdown/datafeed.md:
````diff
@@ -155,6 +155,109 @@ datafeed:
+        await datahose_loop.start()
+```
+
+An example of the usage of the Datahose service can be found [here](../../examples/datafeed/datahose.py )
````
There is an extra space just before `)`

symphony-elias commented on docsrc/markdown/getting_started.md:
````diff
@@ -104,7 +104,7 @@ except KeyboardInterrupt:
 ```
 1. The `SymphonyBdk` class acts as an entry point into the library and provides an API to access
-to the main BDK features such as [Datafeed](./datafeed.md) or [message service](./message_service.md).
+to the main BDK features such as [Datafeed](./datafeed.md), [Datahose](./datafeed.md) or [message service](./message_service.md).
````
It would be better to specify anchors for datafeed and datahose

symphony-elias commented on symphony/bdk/core/service/datafeed/datafeed_loop_v2.py:
```diff
@@ -117,6 +86,8 @@ async def _retrieve_datafeed(self) -> Optional[V5Datafeed]:
                                                            key_manager_token=key_manager_token,
                                                            tag=self._tag)
+
+        datafeeds = list(filter(lambda df: re.compile(DATAFEED_V2_ID_PATTERN).match(df.id), datafeeds))
```
`DATAFEED_V2_ID_PATTERN` could be directly be assigned to `re.compile("^[^\\s_]+_f_[^\\s_]+$")` as compilation is costly (at least more than the matching)

symphony-soufiane committed:
Corrected documentation links + compile regex only once in datafeed_loop_v2

symphony-soufiane committed:
Corrected documentation links + compile regex only once in datafeed_loop_v2

symphony-youri commented on docsrc/markdown/datafeed.md:
````diff
@@ -183,7 +183,7 @@ async def run():
 ```
 
-An example of the usage of the Datahose service can be found [here](../../examples/datafeed/datahose.py )
+An example of the usage of the Datahose service can be found [here](https://github.com/finos/symphony-bdk-python/tree/main/examples/datafeed/datahose.py)
````
can't we keep relative links?

symphony-soufiane replied:
I don't think we can do it since this document will be open on github page which is different from the repo:  https://symphony-bdk-python.finos.org/markdown/datafeed.html

symphony-youri commented on docsrc/markdown/datafeed.md:
```diff
@@ -155,6 +155,109 @@ datafeed:
+Datahose is very similar to datafeed: it enables a bot to receive [_Real Time
+Events_](https://docs.developers.symphony.com/building-bots-on-symphony/datafeed/real-time-events) with the main
+difference that *all* events of the pod are received. The datahose loop is a core service built on top of the events API
+and provide a dedicated contract to bot developers to work with datahose. This is compatible with agent version 20.15 onwards.
```
```suggestion
and provide a dedicated contract to bot developers to work with Datahose. This is compatible with Agent version 20.15 onwards.
```

in the doc (like https://docs.developers.symphony.com/building-bots-on-symphony/datafeed), Datafeed, Agent, and so are capitalized

symphony-soufiane committed:
Make recreate_datafeed and prepare_datafeed not abstract in abstract_datafeed_loop.
Datahose extends abstract_datafeed_loop but it should not implement those methods as they are specific to DF1 and DF2. They are not abstarct methods anymore and they are declared and implemented in the concrete classes for DFV1 and DFV2 loops.

symphony-soufiane committed:
Update docsrc/markdown/datafeed.md

Co-authored-by: Youri Bonnaffé <63661676+symphony-youri@users.noreply.github.com>
